### PR TITLE
Fix Position of Search Icon in Site Creation - Choose a Domain for RTL languages

### LINKED ISSUE #13607: Search icon appears on left side in RTL language
### Expected behavior
The search icon should appear on the right side. Found in https://github.com/wordpress-mobile/WordPress-iOS/pull/13597#discussion_r389808797

### Actual behavior
The search icon appears on the left side 

### Steps to reproduce the behavior
1. Set your device to RTL language 
2. Create a new site
3. See in vertical search and domain search the search icon appears on the left side

##### Tested on [device], iOS [version], WPiOS [version]
iPhone / Simulator 11 pro iOS 13.3 WPiOS 14.4 Hebrew language

### PR DESCRIPTION
**Pull request by ovitrif**
# 📱 Fix Position of Search Icon in RTL languages

Fixes [#13607 Search icon appears on left side in RTL language](https://github.com/wordpress-mobile/WordPress-iOS/issues/13607)

This pull-request fixes the input field in **Site Creation &rarr; Choose a Domain** for RTL languages:
- Position of search / loading icon
- Position of clear button
- Text alignment of the editable part in the text field

I took inspiration from the default iOS layout for search fields in RTL languages:

expected_icon_rtl

## To Test

1. Make sure the language of the device is set to a right to left (RTL) language (Arabic, Hebrew, etc)
If building the app from this pull request in Xcode, you can go to Product → Scheme → Edit Scheme and set the app language to Arabic or Hebrew before building.
2. Start creating a new site using the app (From My Site screen tap the ▼ icon next to the site title)
3. On the Site Picker screen tap on the ➕ menu button
4. Select **Create WordPress.com site** (first option)
5. Choose any design or tap the **SKIP** menu button (doesn't matter for this test)
6. **Verify**
   1. Search icon is positioned at the right of the text
   2. Input placeholder text is aligned to the right
7. Type  anything random for the domain, like `example test`
8. **Verify** 
   1. Loading icon appears at the right of the text (where the search icon was)
   2. After loading results, the search icon is positioned at the right
   3. The text that you entered is aligned to the right
   4. The clear button is positioned at the left of the text input

## Previews

| Before | After |
| ------ | ------ |
| before | after |

## Regression Notes
1. **Potential unintended areas of impact**
    - Layout of the text field component in LTR languages
    - Other places where the text field component is used (I only tested Site Creation → Choose a Domain)
    
2. **What I did to test those areas of impact (or what existing automated tests I relied on)**
✅ Manual testing of the screen in LTR and RTL languages.

3. **What automated tests I added (or what prevented me from doing so)**
I'm not sure if it's applicable to add UI tests for this, I didn't see any in a previous pull request for this input (I.e. : [#13597](https://github.com/wordpress-mobile/WordPress-iOS/pull/13597)), but I'm open to suggestions!

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by twstokes** [comment]
>If building the app from this pull request in Xcode, you can go to Product → Schemes → Edit Scheme and set the app language to Arabic or Hebrew before building.

Whoa, nice! Thanks for sharing this trick. 🙇

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift [hunk 1]
```diff
@@ -105,21 +116,13 @@ final class SearchTextField: UITextField {
...
 
     private func setIconImage(view: UIView) {
-        if traitCollection.layoutDirection == .rightToLeft {
-            rightView = view
-            rightViewMode = .always
-        } else {
-            leftView = view
-            leftViewMode = .always
-        }
+        // Since the RTL layout is already handled elsewhere updating leftVIew is enough here
```
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Tiniest of nits 😄 

```suggestion
        // Since the RTL layout is already handled elsewhere updating leftView is enough here
```

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift [hunk 2]
```diff
@@ -78,7 +89,7 @@ final class SearchTextField: UITextField {
 
     private lazy var searchIconImageView: UIImageView = {
         let iconSize = CGSize(width: Constants.iconDimension, height: Constants.iconDimension)
-        let loupeIcon = UIImage.gridicon(.search, size: iconSize).imageWithTintColor(.listIcon)?.imageFlippedForRightToLeftLayoutDirection()
+        let loupeIcon = UIImage.gridicon(.search, size: iconSize).imageWithTintColor(.listIcon)
```
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift:88]: Makes sense! I think in your original "After" preview the search icon was leaning away from text, but in my testing it was always leaning into it, so I suspect that screenshot may have come before this code change.

image

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift [hunk 3]
```diff
@@ -105,21 +116,13 @@ final class SearchTextField: UITextField {
...
-        } else {
-            leftView = view
-            leftViewMode = .always
-        }
+        // Since the RTL layout is already handled elsewhere updating leftVIew is enough here
+        leftView = view
+        leftViewMode = .always
     }
 
     override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
```
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Should we wipe this overridden function now that we don't do any work in it?

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift [hunk 4]
```diff
@@ -69,6 +76,10 @@ final class SearchTextField: UITextField {
         autocorrectionType = .no
         adjustsFontForContentSizeCategory = true
 
+        if effectiveUserInterfaceLayoutDirection == .rightToLeft {
+            textAlignment = .right
```
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: This piqued my interest because I initially thought we'd want to go with the `natural` text alignment setting, but that's dependent on the keyboard used, so this makes sense!

**Review by twstokes — COMMENTED** [review]
Great work @ovitrif! I made just a couple minor suggestions.

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 1 above)
Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Thank you @twstokes! I used the "Commit suggestion" feature, to learn more of GitHub's features 🙃

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 2 above)
Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift:88]: > Makes sense! I think in your original "After" preview the search icon was leaning away from text, but in my testing it was always leaning into it, so I suspect that screenshot may have come before this code change.

Good observation, Thank you! I got confused after switching back and forth from RTL to LTR 🙈!

I decided to remove the flipping after preparing the pull request, because I noticed in these two cases there is no flipping for RTL languages:
- In the WordPress Android app
- In the system iOS search input

I updated the screenshot to remove the confusion 🙇‍♂️

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 3 above)
Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Good point Tanner!
Thank you for this, I must've missed the fact that there's no more custom logic in this `override`, it's great to get your support on this with a fresh perspective 🙇‍♂️

I'll remove the effectively empty override and push the commit.

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 4 above)
Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Thank you Tanner for opening the discussion here 🙇‍♂️

The fact that the `natural` iOS text alignment is based on the keyboard is new to me, I made this change because I was thinking I'm fixing something that doesn't work as expected, but I see it actually works 'natural' if I use a RTL language for the keyboard.

Now it makes me think that I actually looked at the problem from the perspective of a developer fixing an issue for languages he's not using, while in fact for actual users of RTL languages this might feel natural.

TL; DR I didn't have the right context.

Do you think we should keep this change?

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 3 above)
Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Removed it 👍

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 4 above)
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Your implementation feels right to me based on the default iOS layout for search fields in RTL languages in your description, but perhaps we should consult with a RTL user and get their input to make sure?

**Review by twstokes — APPROVED** [review]
LGTM @ovitrif!

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 4 above)
Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: @yaelirub Pinging you here because you registered the original issue #13607 and I think you could help!

We'd like to get feedback on this from someone with experience in using RTL languages on an iPhone.

In short: We're not 💯 sure if it's ok to always switch the text alignment to right for RTL languages.

Given that we're used to always have a LTR keyboard it took some time to figure out that the text alignment is naturally switched to the right when the keyboard is in a RTL language.

I tested with a RTL language after our discussion here and it doesn't look bad if this change is removed and everything else stays the same in the pull request.

Please let me know if you checked this comment but you feel you won't have time to look into it this week 🙇

Comment (twstokes) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: 👋 Here are the options visually, in case it's helpful:

| Option A | Option B |
| -------- | --------- |
| Option A | Option B |

With Option A the text alignment will always match the app localization independently of the keyboard chosen. Option B will align the text for the localization of the individual keyboard (e.g. English will left align, Hebrew will right align).

Comment (yaelirub) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Thanks for the ping @ovitrif . Option B makes more sense to me

Comment (ovitrif) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Thank you Yael 🙇 

@twstokes I removed the part that was forcing the right text alignment regardless of the keyboard.

This is how it behaves now:

| English Keyboard | Hebrew Keyboard |
| --- | --- |
| ltr_keyboard | rtl_keyboard |

Please do have a look at it when you have time, I'll enable auto-merge to continue if it's all ok 🙇

**Review by twstokes — APPROVED** [review]

File: WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift (hunk 4 above)
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Site Creation/Shared/TitleSubtitleTextfieldHeader.swift]: Thanks @ovitrif! I've approved the PR - these screenshots appear how I'd expect them to after recent discussions. 🚀

**Comment by brcavallaro** [comment]
Hey @ovitrif! When checking and testing the changes from release WPiOS 19.4.0.2, I noticed that the behavior of RTL languages and keyboards is kinda different than what was [discussed in the previous comments from this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/18006#discussion_r814957065).

Once you have a chance, could you please take a look at the videos below? It seems this is still using always the localization text positioning, over the keyboard one. I performed two scenarios, one using Arabic localization + Arabic and English keyboards, and another one, using English localization (non-RTL) + Arabic and English keyboards.

I just wanted to confirm if this is the expected behavior or a regression issue. If it's a regression one, I can submit a new issue for further investigation.

<div align="center">

https://user-images.githubusercontent.com/2813972/159177030-4c6d1081-9587-4903-b11b-2e978ee3b7d3.mov

</div>

<div align="center">

https://user-images.githubusercontent.com/2813972/159177019-70b8130f-9d12-43ea-9817-332355a7f1f3.mov

</div>

Thank you very much!

**Comment by twstokes** [comment]
👋 @brcavallaro! That's interesting. I've only tested English localization, but my input will switch alignment depending on the keyboard chosen. It's normal for the alignment to stay fixed if characters have already been typed, but I can see from your video that the input field appears empty (no "x" button is showing). I'm running 19.4.0.3 from TestFlight on iOS 15.4.

https://user-images.githubusercontent.com/2092798/159178949-c0399b16-258f-4ec7-a85d-e11abddd3415.MP4

**Comment by ovitrif** [comment]
> Hey @ovitrif! When checking and testing the changes from release WPiOS 19.4.0.2, I noticed that the behavior of RTL languages and keyboards is kinda different than what was [discussed in the previous comments from this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/18006#discussion_r814957065).
> 
> Once you have a chance, could you please take a look at the videos below? It seems this is still using always the localization text positioning, over the keyboard one. I performed two scenarios, one using Arabic localization + Arabic and English keyboards, and another one, using English localization (non-RTL) + Arabic and English keyboards.

Hey @brcavallaro 👋

Thank you for testing the fix and reporting this issue you are encountering.

I tried to reproduce the issue you reported on `19.4`* but I could only get the same behavior Tanner already mentioned in his [comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/18006#issuecomment-1073315314).

<details>
<summary><h2>Previews</h2></summary>

### Phone in English Language
https://user-images.githubusercontent.com/4588074/159275416-a9a6e394-86a7-4285-9d51-008a32bcee37.MP4

### Phone in Arabic language
https://user-images.githubusercontent.com/4588074/159275587-387ca9df-4549-4cbe-a07d-137b149fe525.MP4
</details>

---

**\*** WordPress iOS 19.4 (19.4.0.2) from TestFlight, iPhone 13 Pro Max running iOS 15.4

---

> I just wanted to confirm if this is the expected behavior or a regression issue. If it's a regression one, I can submit a new issue for further investigation.

Since we're not getting the behaviour you're experiencing — and what you're experiencing is not the expected behaviour — I think it's ok to submit a new issue, where ideally you add a bit more details about your setup (iPhone variant, App version). Hopefully we'll be able to reproduce and proceed to find a fix 🤞🏽

**Comment by brcavallaro** [comment]
Hey @ovitrif and @twstokes! 👋

Thanks for taking a look at the comment and replying here! I'll raise a new issue here with more details about the OS/device and keep you posted about it once that's done.

**Update:** Issue https://github.com/wordpress-mobile/WordPress-iOS/issues/18178 created.
